akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -512,6 +520,20 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
         false
     }
 
+  //this limits number of messages that can be processed synchronously during one actor receive.
+  private val shortCircuitLimit: Int = _initial.mat.settings.syncProcessingLimit
+  private var shortCircuitBuffer: FixedSizeBuffer[BoundaryEvent] = null
+  private var currentLimit: Int = shortCircuitLimit
+
+  def enqueueToShortCircuit(input: AsyncInput): Boolean = {
+    if (shortCircuitBuffer == null) shortCircuitBuffer = FixedSizeBuffer[BoundaryEvent](shortCircuitLimit)
+    if (!shortCircuitBuffer.isFull && currentLimit > 0) {
```

Comment:
> While processing the inputs from the list it is random chance whether any given callback will be enqueued or not—in particular it is possible that after having sent one callback via the ActorRef the next one will go to the short circuit.

It will go short circuit based on two conditions:
1) AsyncInput is for shell that locates in the same ActorGraphInterpreter (same actor)
2) limit of AsyncInput in buffer (per ActorGraphInterpreter.received) is not reached

still there is a chance that buffer was full in one moment of time so this AsyncInput went via ActorRef and next go via short circuit. This can be fixed with some flag or having unbounded queue for short circuit